### PR TITLE
[1994] Set up allocation subject + subject specialisms

### PR DESCRIPTION
**Pull request by JonF1**
### Context
https://trello.com/c/pXx6En7b/1994-s-add-allocation-subjects-and-specialism-subjects

### Changes proposed in this pull request
Add subject specialism and allocation subject model + migrations + associations with fields specified in ticket
Model tests to verify associations

### CODE REVIEW

**Comment by JonF1** [comment]
[[S] Add allocation subjects and specialism subjects](https://trello.com/c/pXx6En7b/1994-s-add-allocation-subjects-and-specialism-subjects)

File: db/migrate/20210616112302_create_subject_specialisms.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateSubjectSpecialisms < ActiveRecord::Migration[6.1]
+  def change
+    create_table :subject_specialisms do |t|
+      t.string :name
+      t.integer :allocation_subject_id
```
Comment (darokel) [inline db/migrate/20210616112302_create_subject_specialisms.rb:7]:
```
t.references :allocation_subject, index: true, null: false, foreign_key: { to_table: :allocation_subjects }
```

Comment (JonF1) [reply db/migrate/20210616112302_create_subject_specialisms.rb]: 🤦

File: db/migrate/20210616110455_create_allocation_subjects.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class CreateAllocationSubjects < ActiveRecord::Migration[6.1]
+  def change
+    create_table :allocation_subjects do |t|
+      t.string :name
```
Comment (darokel) [inline db/migrate/20210616110455_create_allocation_subjects.rb:6]:
```suggestion
t.string :name, null: false, index: { unique: true }
```

File: db/migrate/20210616124359_create_subject_specialisms.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateSubjectSpecialisms < ActiveRecord::Migration[6.1]
+  def change
+    create_table :subject_specialisms do |t|
+      t.string :name
```
Comment (darokel) [inline db/migrate/20210616124359_create_subject_specialisms.rb:6]:
```suggestion
t.string :name, null: false, index: { unique: true }
```

File: db/migrate/20210616124359_create_subject_specialisms.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateSubjectSpecialisms < ActiveRecord::Migration[6.1]
+  def change
+    create_table :subject_specialisms do |t|
+      t.string :name, null: false, index: { unique: true }
+      t.references :allocation_subject, index: true, null: false, foreign_key: { to_table: :allocation_subjects }
```
Comment (jhenderson) [inline db/migrate/20210616124359_create_subject_specialisms.rb:7]:
```suggestion
      t.references :allocation_subject, null: false, foreign_key: true
```

Comment (jhenderson) [reply db/migrate/20210616124359_create_subject_specialisms.rb:7]: `to_table` is only needed if the name of the foreign key was different from the rails convention.

Comment (JonF1) [reply db/migrate/20210616124359_create_subject_specialisms.rb]: fixed

**Review by jhenderson — APPROVED** [review]

**Pull request merged**